### DISCUSSION TITLE
DSL way for building table component

### DISCUSSION THREAD

**Discussion by bykka** [richard-scryber/scryber.core#110] [General]
Hi,
I'm looking a way how to define table component in more declarative way
For how I have the next code
```csharp
public static Component BuildTableComponent(string[][] data)
    {
        var tbl = new TableGrid
        {
            FullWidth = true
        };
        
        data.Select(rowData => rowData
                .Select(colData => new TableCell
                {
                    Contents = { colData }
                })
                .Aggregate(new TableRow(), (tableRow, cell) =>
                {
                    tableRow.Cells.Add(cell);
                    return tableRow;
                }))
            .Aggregate(tbl, (grid, row) =>
            {
                grid.Rows.Add(row);
                return grid;
            });
        
        return tbl;
    }
```
As you can see is not pleased to read. Especially because of nested Aggregate methods.
It happens because properties Rows, Cell, Contents do not allow to set values
If to assume that I can set a new value for the those properties than code could look like the next:
```csharp
public static Component TableWidgetComponent(string[][] data)
    {
        return new TableGrid
        {
            FullWidth = true,
            Rows = data.Select(rowData => new TableRow
            {
                Cells = rowData.Select(colData => new TableCell
                {
                    Contents = { colData }
                })
            })
        };
    }
```
Maybe there is a some another way for defining document/component in more declarative way which I have missed?

**Comment by richard-scryber** [comment]
Hi Pavlo, as the TableRowList and TableCellList are directly associated with the Table and Row instances (so that parents can be traversed back up the hierarchy), then I can look at it, but I feel that setters would become exhaustive in checking, updating parent ownership etc.

You could go back to using template content as a string. It then becomes really flexible on the data structure and content...

```
private static TableGrid BuildDynamicTable()
        {
            var table = new Scryber.Components.TableGrid() { FullWidth = true };
            table.DataContent = "<template xmlns='http://www.w3.org/1999/xhtml' data-bind='{{myModel}}'>" +
                "<tr>" +
                    "<template data-bind='{{.}}'>" +
                        "<td><span>{{.}}</span></td>" +
                    "</template>" +
                "</tr>" +
            "</template>";
            return table;
        }
```

Then set the data in your execution

```
string[][] data = GetData();
doc.Params["myModel"] = data;
```


Or just go with old school foreach loops. :-)

**Reply by bykka** [reply]
A set method does not mean that you need to assign a new collection. You can iterate the collection and add elements into existed collection
For example:
```csharp
        public TableCellList Cells
        {
            get 
            {
                if (_cells == null)
                    _cells = new TableCellList(this.InnerContent);

                return _cells;
            }
            set
            {
                foreach (var cell in value)
                {
                    this.Cells.Add(cell);
                }
            }
        }
```

**Reply by richard-scryber** [reply]
Fair shout, but the property type would then have to be IEnumerable<TableCell> as far as I am aware, or IList<TableCell> if you used the ToList() method after select. The TableCellList does not have a parameterless constructor because it uses the inner collection with parent ownership.

However, an AddRange method could be a nice work around allowing Select (and Where, etc)...

```csharp
        private TableGrid GenerateTableGridWithAddRange(string[][] data)
        {

            var grid = new TableGrid() { FullWidth = true };
            grid.Rows.AddRange(data.Select(rowData => {

                    var row = new TableRow();
                    row.Cells.AddRange(rowData.Select(cellData => {

                        var cell = new TableCell();
                        cell.Contents.Add(cellData);

                        return cell;
                    }));

                    return row;
                }));
            return grid;
        }
```

This is easy to implement on the base class and will be available to Document.Pages, List.Items, Table.Rows, Row.Cells, and all other Contents?
